src/OrchardCore.Modules/OrchardCore.Navigation/PagerShapes.cs

```diff
@@ -398,14 +373,15 @@ public Task<IHtmlContent> Pager(Shape shape, DisplayContext displayContext)
             shape.Metadata.Alternates.Clear();
             shape.Metadata.Type = "List";
 
-            var routeData = new RouteValueDictionary(Html.ViewContext.RouteData.Values);
+
+            var routeData = GetRouteData(shape, displayContext, Html);
 
             // Allows to pass custom url params to PagerSlim
             if (UrlParams != null)
             {
                 foreach (var item in UrlParams)
                 {
-                    routeData.Add(item.Key, item.Value);
+                    routeData[item.Key] = item.Value;
```

Comment:
Okay unless we would prefer `routeData.TryAdd()` to check if the key already exists, I'm talking about this just because I saw that in `GetRouteData()` this is what we do when we update route data from the request query.

But not sure at all what would be better to do here, Hmm, before with `.Add()`, it was failing if the key already exists.

Reply:
I've not tested this yet but if it is a dictionary I'm not sure that affecting a key that doesn't exist will create the key/value. This is why I was using the `.Add()` and affecting only if it was existing. These are code optimizations that only will give the same result in the end. I'm trying to fix the issues we have here with the Pagers before anything. We have an SEO issue with pages that can have 2 different URLs as an example. We talked about this in this afternoon's meeting.

Reply:
Okay no problem just because in `GetRouteData()` I saw it doesn't do the same ;)

`.Add()` adds if not exists but throw if exists.
`TryAdd()` only set if not exists and never throws.
`[key] = value` set or replace existing value and never throws.

In `GetRouteData()` we were using a kind of `TryAdd()` by 1st checking for the key.

Reply:
If I'm doing `.TryAdd()` it will add a value only if it doesn't exist. Though it probably won't update the value if it exists and that is what I want to do. So, I think it `.TryAdd()` is not appropriate.

Reply:
Okay so `[key] = value` is the best, as you already did ;)

Hmm, but then maybe this is what we also want in `GetRouteData()`.